File: paykit-mobile/android-demo/app/src/main/java/com/paykit/demo/storage/PaymentRequestStorage.kt
package com.paykit.demo.storage

import android.content.Context
import android.util.Log
import androidx.security.crypto.EncryptedSharedPreferences
import androidx.security.crypto.MasterKey
import kotlinx.serialization.Serializable
import kotlinx.serialization.encodeToString
import kotlinx.serialization.json.Json
import java.util.Date

/**
 * Manages persistent storage of payment requests using EncryptedSharedPreferences.
 */
class PaymentRequestStorage(context: Context, private val identityName: String) {
    
    companion object {
        private const val TAG = "PaymentRequestStorage"
        private const val MAX_REQUESTS_TO_KEEP = 200
    }
    
    private val PREFS_NAME = "paykit_payment_requests_$identityName"
    private val REQUESTS_KEY = "requests_list"
    
    private val json = Json { 
        ignoreUnknownKeys = true 
        encodeDefaults = true
    }
    
    private val prefs by lazy {
        try {
            val masterKey = MasterKey.Builder(context)
                .setKeyScheme(MasterKey.KeyScheme.AES256_GCM)
                .build()
            
            EncryptedSharedPreferences.create(
                context,
                PREFS_NAME,
                masterKey,
                EncryptedSharedPreferences.PrefKeyEncryptionScheme.AES256_SIV,
                EncryptedSharedPreferences.PrefValueEncryptionScheme.AES256_GCM
            )
        } catch (e: Exception) {
            Log.e(TAG, "Failed to create encrypted prefs, falling back to regular prefs", e)
            context.getSharedPreferences(PREFS_NAME, Context.MODE_PRIVATE)
        }
    }
    
    // In-memory cache
    private var requestsCache: List<StoredPaymentRequest>? = null
    
    // MARK: - CRUD Operations
    
    /**
     * Get all requests (newest first)
     */
    fun listRequests(): List<StoredPaymentRequest> {
        requestsCache?.let { return it }
        
        return try {
            val jsonString = prefs.getString(REQUESTS_KEY, null) ?: return emptyList()
            val requests = json.decodeFromString<List<StoredPaymentRequest>>(jsonString)
                .sortedByDescending { it.createdAt }
            requestsCache = requests
            requests
        } catch (e: Exception) {
            Log.e(TAG, "Failed to load requests: ${e.message}")
            emptyList()
        }
    }
    
    /**
     * Get pending requests only
     */
    fun pendingRequests(): List<StoredPaymentRequest> {
        return listRequests().filter { it.status == PaymentRequestStatus.PENDING }
    }
    
    /**
     * Get requests filtered by status
     */
    fun listRequests(status: PaymentRequestStatus): List<StoredPaymentRequest> {
        return listRequests().filter { it.status == status }
    }
    
    /**
     * Get requests filtered by direction
     */
    fun listRequests(direction: RequestDirection): List<StoredPaymentRequest> {
        return listRequests().filter { it.direction == direction }
    }
    
    /**
     * Get recent requests (limited count)
     */
    fun recentRequests(limit: Int = 10): List<StoredPaymentRequest> {
        return listRequests().take(limit)
    }
    
    /**
     * Get a specific request
     */
    fun getRequest(id: String): StoredPaymentRequest? {
        return listRequests().find { it.id == id }
    }
    
    /**
     * Add a new request
     */
    fun addRequest(request: StoredPaymentRequest) {
        val requests = listRequests().toMutableList()
        
        // Add new request at the beginning (newest first)
        requests.add(0, request)
        
        // Trim to max size
        val trimmed = if (requests.size > MAX_REQUESTS_TO_KEEP) {
            requests.take(MAX_REQUESTS_TO_KEEP)
        } else {
            requests
        }
        
        persistRequests(trimmed)
    }
    
    /**
     * Update an existing request
     */
    fun updateRequest(request: StoredPaymentRequest) {
        val requests = listRequests().toMutableList()
        val index = requests.indexOfFirst { it.id == request.id }
        
        if (index >= 0) {
            requests[index] = request
            persistRequests(requests)
        }
    }
    
    /**
     * Update request status
     */
    fun updateStatus(id: String, status: PaymentRequestStatus) {
        val request = getRequest(id) ?: return
        updateRequest(request.copy(status = status))
    }
    
    /**
     * Delete a request
     */
    fun deleteRequest(id: String) {
        val requests = listRequests().toMutableList()
        requests.removeAll { it.id == id }
        persistRequests(requests)
    }
    
    /**
     * Check and mark expired requests
     */
    fun checkExpirations() {
        val now = System.currentTimeMillis()
        val requests = listRequests().toMutableList()
        var hasChanges = false
        
        for (i in requests.indices) {
            val request = requests[i]
            if (request.status == PaymentRequestStatus.PENDING && 
                request.expiresAt != null && 
                request.expiresAt < now) {
                requests[i] = request.copy(status = PaymentRequestStatus.EXPIRED)
                hasChanges = true
            }
        }
        
        if (hasChanges) {
            persistRequests(requests)
        }
    }
    
    /**
     * Clear all requests
     */
    fun clearAll() {
        persistRequests(emptyList())
    }
    
    // MARK: - Statistics
    
    /**
     * Count of pending requests
     */
    fun pendingCount(): Int {
        return listRequests(PaymentRequestStatus.PENDING).size
    }
    
    /**
     * Count of incoming pending requests
     */
    fun incomingPendingCount(): Int {
        return listRequests(RequestDirection.INCOMING).count { it.status == PaymentRequestStatus.PENDING }
    }
    
    /**
     * Count of outgoing pending requests
     */
    fun outgoingPendingCount(): Int {
        return listRequests(RequestDirection.OUTGOING).count { it.status == PaymentRequestStatus.PENDING }
    }
    
    // MARK: - Private
    
    private fun persistRequests(requests: List<StoredPaymentRequest>) {
        try {
            val jsonString = json.encodeToString(requests)
            prefs.edit().putString(REQUESTS_KEY, jsonString).apply()
            requestsCache = requests
        } catch (e: Exception) {
            Log.e(TAG, "Failed to save requests: ${e.message}")
        }
    }
}

// MARK: - Data Models

/**
 * A payment request stored in persistent storage
 */
@Serializable
data class StoredPaymentRequest(
    val id: String,
    val fromPubkey: String,
    val toPubkey: String,
    val amountSats: Long,
    val currency: String,
    val methodId: String,
    val description: String,
    val createdAt: Long,
    val expiresAt: Long?,
    val status: PaymentRequestStatus,
    val direction: RequestDirection
) {
    /**
     * Display name for the counterparty
     */
    val counterpartyName: String
        get() {
            // In a real app, this would look up the contact name
            val key = if (direction == RequestDirection.INCOMING) fromPubkey else toPubkey
            return if (key.length > 12) {
                "${key.take(6)}...${key.takeLast(4)}"
            } else {
                key
            }
        }
    
    /**
     * Get createdAt as Date
     */
    val createdAtDate: Date
        get() = Date(createdAt)
    
    /**
     * Get expiresAt as Date
     */
    val expiresAtDate: Date?
        get() = expiresAt?.let { Date(it) }
    
    companion object {
        /**
         * Create from FFI PaymentRequest
         */
        fun fromFFI(
            ffiRequest: com.paykit.mobile.PaymentRequest, 
            direction: RequestDirection
        ): StoredPaymentRequest {
            return StoredPaymentRequest(
                id = ffiRequest.requestId,
                fromPubkey = ffiRequest.fromPubkey,
                toPubkey = ffiRequest.toPubkey,
                amountSats = ffiRequest.amountSats,
                currency = ffiRequest.currency,
                methodId = ffiRequest.methodId,
                description = ffiRequest.description,
                createdAt = ffiRequest.createdAt * 1000, // Convert to milliseconds
                expiresAt = ffiRequest.expiresAt?.let { it * 1000 },
                status = PaymentRequestStatus.PENDING,
                direction = direction
            )
        }
    }
}

/**
 * Status of a payment request
 */
@Serializable
enum class PaymentRequestStatus {
    PENDING,
    ACCEPTED,
    DECLINED,
    EXPIRED,
    PAID;
    
    val displayName: String
        get() = when (this) {
            PENDING -> "Pending"
            ACCEPTED -> "Accepted"
            DECLINED -> "Declined"
            EXPIRED -> "Expired"
            PAID -> "Paid"
        }
}

/**
 * Direction of the request (incoming = someone is requesting from you)
 */
@Serializable
enum class RequestDirection {
    INCOMING,  // Someone is requesting payment from you
    OUTGOING   // You are requesting payment from someone
}

File: paykit-mobile/ios-demo/PaykitDemo/PaykitDemo/PaykitDemo/Storage/PaymentRequestStorage.swift
//
//  PaymentRequestStorage.swift
//  PaykitDemo
//
//  Persistent storage for payment requests using Keychain.
//

import Foundation

/// Manages persistent storage of payment requests
class PaymentRequestStorage {
    
    private let keychain: KeychainStorage
    private let requestsKey = "paykit.payment_requests.list"
    private let maxRequestsToKeep = 200  // Limit stored requests
    
    // In-memory cache
    private var requestsCache: [StoredPaymentRequest]?
    
    init(keychain: KeychainStorage = KeychainStorage(serviceIdentifier: "com.paykit.demo")) {
        self.keychain = keychain
    }
    
    // MARK: - CRUD Operations
    
    /// Get all requests (newest first)
    func listRequests() -> [StoredPaymentRequest] {
        if let cached = requestsCache {
            return cached
        }
        
        do {
            guard let data = try keychain.retrieve(key: requestsKey) else {
                return []
            }
            var requests = try JSONDecoder().decode([StoredPaymentRequest].self, from: data)
            // Sort by date, newest first
            requests.sort { $0.createdAt > $1.createdAt }
            requestsCache = requests
            return requests
        } catch {
            print("PaymentRequestStorage: Failed to load requests: \(error)")
            return []
        }
    }
    
    /// Get pending requests only
    func pendingRequests() -> [StoredPaymentRequest] {
        return listRequests().filter { $0.status == .pending }
    }
    
    /// Get requests filtered by status
    func listRequests(status: PaymentRequestStatus) -> [StoredPaymentRequest] {
        return listRequests().filter { $0.status == status }
    }
    
    /// Get requests filtered by direction
    func listRequests(direction: RequestDirection) -> [StoredPaymentRequest] {
        return listRequests().filter { $0.direction == direction }
    }
    
    /// Get recent requests (limited count)
    func recentRequests(limit: Int = 10) -> [StoredPaymentRequest] {
        return Array(listRequests().prefix(limit))
    }
    
    /// Get a specific request
    func getRequest(id: String) -> StoredPaymentRequest? {
        return listRequests().first { $0.id == id }
    }
    
    /// Add a new request
    func addRequest(_ request: StoredPaymentRequest) throws {
        var requests = listRequests()
        
        // Add new request at the beginning (newest first)
        requests.insert(request, at: 0)
        
        // Trim to max size
        if requests.count > maxRequestsToKeep {
            requests = Array(requests.prefix(maxRequestsToKeep))
        }
        
        try persistRequests(requests)
    }
    
    /// Update an existing request
    func updateRequest(_ request: StoredPaymentRequest) throws {
        var requests = listRequests()
        
        guard let index = requests.firstIndex(where: { $0.id == request.id }) else {
            throw PaymentRequestStorageError.notFound(id: request.id)
        }
        
        requests[index] = request
        try persistRequests(requests)
    }
    
    /// Update request status
    func updateStatus(id: String, status: PaymentRequestStatus) throws {
        guard var request = getRequest(id: id) else {
            throw PaymentRequestStorageError.notFound(id: id)
        }
        request.status = status
        try updateRequest(request)
    }
    
    /// Delete a request
    func deleteRequest(id: String) throws {
        var requests = listRequests()
        requests.removeAll { $0.id == id }
        try persistRequests(requests)
    }
    
    /// Check and mark expired requests
    func checkExpirations() throws {
        let now = Date()
        var requests = listRequests()
        var hasChanges = false
        
        for i in 0..<requests.count {
            if requests[i].status == .pending,
               let expiresAt = requests[i].expiresAt,
               expiresAt < now {
                requests[i].status = .expired
                hasChanges = true
            }
        }
        
        if hasChanges {
            try persistRequests(requests)
        }
    }
    
    /// Clear all requests
    func clearAll() throws {
        try persistRequests([])
    }
    
    // MARK: - Statistics
    
    /// Count of pending requests
    func pendingCount() -> Int {
        return listRequests(status: .pending).count
    }
    
    /// Count of incoming pending requests
    func incomingPendingCount() -> Int {
        return listRequests(direction: .incoming).filter { $0.status == .pending }.count
    }
    
    /// Count of outgoing pending requests
    func outgoingPendingCount() -> Int {
        return listRequests(direction: .outgoing).filter { $0.status == .pending }.count
    }
    
    // MARK: - Private
    
    private func persistRequests(_ requests: [StoredPaymentRequest]) throws {
        let data = try JSONEncoder().encode(requests)
        try keychain.store(key: requestsKey, data: data)
        requestsCache = requests
    }
}

// MARK: - Data Models

/// A payment request stored in persistent storage
struct StoredPaymentRequest: Identifiable, Codable {
    let id: String
    let fromPubkey: String
    let toPubkey: String
    let amountSats: Int64
    let currency: String
    let methodId: String
    let description: String
    let createdAt: Date
    let expiresAt: Date?
    var status: PaymentRequestStatus
    let direction: RequestDirection
    
    /// Display name for the counterparty
    var counterpartyName: String {
        // In a real app, this would look up the contact name
        let key = direction == .incoming ? fromPubkey : toPubkey
        if key.count > 12 {
            return String(key.prefix(6)) + "..." + String(key.suffix(4))
        }
        return key
    }
    
    /// Create from FFI PaymentRequest
    static func fromFFI(_ ffiRequest: PaymentRequest, direction: RequestDirection) -> StoredPaymentRequest {
        StoredPaymentRequest(
            id: ffiRequest.requestId,
            fromPubkey: ffiRequest.fromPubkey,
            toPubkey: ffiRequest.toPubkey,
            amountSats: ffiRequest.amountSats,
            currency: ffiRequest.currency,
            methodId: ffiRequest.methodId,
            description: ffiRequest.description,
            createdAt: Date(timeIntervalSince1970: Double(ffiRequest.createdAt)),
            expiresAt: ffiRequest.expiresAt.map { Date(timeIntervalSince1970: Double($0)) },
            status: .pending,
            direction: direction
        )
    }
}

/// Status of a payment request
enum PaymentRequestStatus: String, Codable {
    case pending = "Pending"
    case accepted = "Accepted"
    case declined = "Declined"
    case expired = "Expired"
    case paid = "Paid"
}

/// Direction of the request (incoming = someone is requesting from you)
enum RequestDirection: String, Codable {
    case incoming  // Someone is requesting payment from you
    case outgoing  // You are requesting payment from someone
}

// MARK: - Errors

enum PaymentRequestStorageError: LocalizedError {
    case encodingFailed
    case decodingFailed
    case notFound(id: String)
    
    var errorDescription: String? {
        switch self {
        case .encodingFailed:
            return "Failed to encode payment requests"
        case .decodingFailed:
            return "Failed to decode payment requests"
        case .notFound(let id):
            return "Payment request not found: \(id)"
        }
    }
}

